credoai/modules/dataset_modules/dataset_base.py

```diff
@@ -61,4 +65,86 @@ def _make_pipe(self):
                [('scaler', StandardScaler()),
                 ('model', model)])
         return pipe
+
+    
+    def _find_categorical_features(self, df):
+        """Estimates categorical features
+        Logic: all non-float-type columns are categorical
+
+        Parameters
+        ----------
+        df : pandas.dataframe
+            A dataframe
+
+        Returns
+        -------
+        list
+            Names of categorical features
+        """        
+        cols = df.columns
+        float_cols = df.select_dtypes(include=[np.float]).columns
+        cat_cols = list(set(cols) - set(float_cols))
+        return cat_cols
+
+    def _calculate_mutual_information(self, categorical_features=None, normalize=True):
+        """Calculates normalized mutual information between sensitive feature and other features
+        Mutual information is the "amount of information" obtained about the sensitive feature by observing the other feature.
+        It can therefore be used to proxy detection purposes.
+
+        Parameters
+        ----------
+        categorical_features : [str], optional
+            List of the categorical features (including the sensitive attribute) in the dataset, by default None
+            If not provided, all non-float-type features are considered as categorical features
+
+        Returns
+        -------
+        dict
+            Normalized mutual information between sensitive feature and features and their considered feature type (categorical/continuous)
+            Normalized mutual information between sensitive feature and itself is always 1, but is included to report its considered type
+        """        
+    
+        # Create a pandas dataframe of features and sensitive feature
+        if isinstance(self.sensitive_features, pd.Series):
+            df = pd.concat([self.X, self.sensitive_features], axis=1)
+            sensitive_feature_name = self.sensitive_features.name
+        else:
+            df = self.X.copy()
+            df['sensitive_feature'] = sensitive_features
+            sensitive_feature_name = 'sensitive_feature'
+
+        # Estimate categorical features, if not provided
+        if not categorical_features:
+            categorical_features = self._find_categorical_features(df)
+
+        # Encode categorical features
+        for col in categorical_features:
+            df[col] = df[col].astype('category').cat.codes
+
+        discrete_features = [True if col in categorical_features else False for col in df.columns]
+
+        # Use the right mutual information methods based on the feature type of the sensitive attribute
+        if sensitive_feature_name in categorical_features:
```

Comment:
This is making me think we may want to have an overall randomization seed for lens that propagates through. No need to worry about that now though!

Reply:
Sounds good!